commit/merkleroot/rmn/controller.go

```diff
@@ -323,14 +347,53 @@ func (c *controller) listenForRmnObservationResponses(
 	}
 }
 
+// gotSufficientObservationResponses checks if we got enough observation responses for each source chain.
+// Enough meaning that we got at least #minObservers observing the same merkle root for a target chain.
+func gotSufficientObservationResponses(
+	updateRequests map[uint64]updateRequestWithMeta,
+	rmnObservationResponses []rmnSignedObservationWithMeta,
+	minObservers int,
+) bool {
+	merkleRootsCount := make(map[uint64]map[cciptypes.Bytes32]int)
+	for _, signedObs := range rmnObservationResponses {
+		for _, lu := range signedObs.SignedObservation.Observation.FixedDestLaneUpdates {
+			if _, exists := merkleRootsCount[lu.LaneSource.SourceChainSelector]; !exists {
+				merkleRootsCount[lu.LaneSource.SourceChainSelector] = make(map[cciptypes.Bytes32]int)
+			}
+			merkleRootsCount[lu.LaneSource.SourceChainSelector][cciptypes.Bytes32(lu.Root)]++
+		}
+	}
+
+	for sourceChain := range updateRequests {
+		// make sure we got at least #minObservers observing the same merkle root for a target chain.
+		countsPerRoot, ok := merkleRootsCount[sourceChain]
+		if !ok || len(countsPerRoot) == 0 {
+			return false
+		}
+
+		values := maps.Values(countsPerRoot)
+		sort.Slice(values, func(i, j int) bool { return values[i] < values[j] })
+		if values[len(values)-1] < minObservers {
+			return false
+		}
+	}
+
+	return true
+}
+
 // nolint:gocyclo // todo
```

Comment:
Didn't notice this `nolint` before, is there any way we can reduce cyclomatic complexity of this func?

Reply:
It's a bit hard actually, i'll try in a follow up.
I enable gocyclo linter recently.